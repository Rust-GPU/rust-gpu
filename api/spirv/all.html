<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="List of all items in this crate"><title>List of all items in this crate</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod sys"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv/index.html">spirv</a><span class="version">0.3.0+sdk-1.3.268.0</span></h2></div><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#constants">Constants</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><h1>List of all items</h1><h3 id="structs">Structs</h3><ul class="all-items"><li><a href="struct.CooperativeMatrixOperands.html">CooperativeMatrixOperands</a></li><li><a href="struct.FPFastMathMode.html">FPFastMathMode</a></li><li><a href="struct.FragmentShadingRate.html">FragmentShadingRate</a></li><li><a href="struct.FunctionControl.html">FunctionControl</a></li><li><a href="struct.ImageOperands.html">ImageOperands</a></li><li><a href="struct.KernelProfilingInfo.html">KernelProfilingInfo</a></li><li><a href="struct.LoopControl.html">LoopControl</a></li><li><a href="struct.MemoryAccess.html">MemoryAccess</a></li><li><a href="struct.MemorySemantics.html">MemorySemantics</a></li><li><a href="struct.RayFlags.html">RayFlags</a></li><li><a href="struct.SelectionControl.html">SelectionControl</a></li></ul><h3 id="enums">Enums</h3><ul class="all-items"><li><a href="enum.AccessQualifier.html">AccessQualifier</a></li><li><a href="enum.AddressingModel.html">AddressingModel</a></li><li><a href="enum.BuiltIn.html">BuiltIn</a></li><li><a href="enum.CLOp.html">CLOp</a></li><li><a href="enum.Capability.html">Capability</a></li><li><a href="enum.CooperativeMatrixLayout.html">CooperativeMatrixLayout</a></li><li><a href="enum.CooperativeMatrixUse.html">CooperativeMatrixUse</a></li><li><a href="enum.Decoration.html">Decoration</a></li><li><a href="enum.Dim.html">Dim</a></li><li><a href="enum.ExecutionMode.html">ExecutionMode</a></li><li><a href="enum.ExecutionModel.html">ExecutionModel</a></li><li><a href="enum.FPDenormMode.html">FPDenormMode</a></li><li><a href="enum.FPOperationMode.html">FPOperationMode</a></li><li><a href="enum.FPRoundingMode.html">FPRoundingMode</a></li><li><a href="enum.FunctionParameterAttribute.html">FunctionParameterAttribute</a></li><li><a href="enum.GLOp.html">GLOp</a></li><li><a href="enum.GroupOperation.html">GroupOperation</a></li><li><a href="enum.HostAccessQualifier.html">HostAccessQualifier</a></li><li><a href="enum.ImageChannelDataType.html">ImageChannelDataType</a></li><li><a href="enum.ImageChannelOrder.html">ImageChannelOrder</a></li><li><a href="enum.ImageFormat.html">ImageFormat</a></li><li><a href="enum.InitializationModeQualifier.html">InitializationModeQualifier</a></li><li><a href="enum.KernelEnqueueFlags.html">KernelEnqueueFlags</a></li><li><a href="enum.LinkageType.html">LinkageType</a></li><li><a href="enum.LoadCacheControl.html">LoadCacheControl</a></li><li><a href="enum.MemoryModel.html">MemoryModel</a></li><li><a href="enum.Op.html">Op</a></li><li><a href="enum.OverflowModes.html">OverflowModes</a></li><li><a href="enum.PackedVectorFormat.html">PackedVectorFormat</a></li><li><a href="enum.QuantizationModes.html">QuantizationModes</a></li><li><a href="enum.RayQueryCandidateIntersectionType.html">RayQueryCandidateIntersectionType</a></li><li><a href="enum.RayQueryCommittedIntersectionType.html">RayQueryCommittedIntersectionType</a></li><li><a href="enum.RayQueryIntersection.html">RayQueryIntersection</a></li><li><a href="enum.SamplerAddressingMode.html">SamplerAddressingMode</a></li><li><a href="enum.SamplerFilterMode.html">SamplerFilterMode</a></li><li><a href="enum.Scope.html">Scope</a></li><li><a href="enum.SourceLanguage.html">SourceLanguage</a></li><li><a href="enum.StorageClass.html">StorageClass</a></li><li><a href="enum.StoreCacheControl.html">StoreCacheControl</a></li></ul><h3 id="types">Type Aliases</h3><ul class="all-items"><li><a href="type.Word.html">Word</a></li></ul><h3 id="constants">Constants</h3><ul class="all-items"><li><a href="constant.MAGIC_NUMBER.html">MAGIC_NUMBER</a></li><li><a href="constant.MAJOR_VERSION.html">MAJOR_VERSION</a></li><li><a href="constant.MINOR_VERSION.html">MINOR_VERSION</a></li><li><a href="constant.REVISION.html">REVISION</a></li></ul></section></div></main></body></html>