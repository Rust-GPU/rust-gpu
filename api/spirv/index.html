<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="The SPIR-V header."><title>spirv - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-46f98efaafac5295.ttf.woff2,FiraSans-Regular-018c141bf0843ffd.woff2,FiraSans-Medium-8f9a781e4970d388.woff2,SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2,SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-081576b923113409.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="spirv" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (244da22fa 2024-04-23)" data-channel="nightly" data-search-js="search-bf21c90c8c1d92b1.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../static.files/storage-e32f0c247825364d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-20a3ad099b048cf2.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-09095024cf37855e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../spirv/index.html">spirv</a><span class="version">0.3.0+sdk-1.3.268.0</span></h2></div><div class="sidebar-elems"><ul class="block"><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#constants">Constants</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../spirv/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Type ‘S’ or ‘/’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings">Settings</a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">spirv</a><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><span class="out-of-band"><a class="src" href="../src/spirv/lib.rs.html#1-14">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>The SPIR-V header.</p>
<p>This crate contains Rust definitions of all SPIR-V structs, enums,
and constants.</p>
<p>The version of this crate is the version of SPIR-V it contains.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.CooperativeMatrixOperands.html" title="struct spirv::CooperativeMatrixOperands">CooperativeMatrixOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_operands_a_cooperative_matrix_operands">CooperativeMatrixOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.FPFastMathMode.html" title="struct spirv::FPFastMathMode">FPFastMathMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_fast_math_mode_a_fp_fast_math_mode">FPFastMathMode</a></div></li><li><div class="item-name"><a class="struct" href="struct.FragmentShadingRate.html" title="struct spirv::FragmentShadingRate">FragmentShadingRate</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fragment_shading_rate_a_fragment_shading_rate">FragmentShadingRate</a></div></li><li><div class="item-name"><a class="struct" href="struct.FunctionControl.html" title="struct spirv::FunctionControl">FunctionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_control_a_function_control">FunctionControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.ImageOperands.html" title="struct spirv::ImageOperands">ImageOperands</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_operands_a_image_operands">ImageOperands</a></div></li><li><div class="item-name"><a class="struct" href="struct.KernelProfilingInfo.html" title="struct spirv::KernelProfilingInfo">KernelProfilingInfo</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_profiling_info_a_kernel_profiling_info">KernelProfilingInfo</a></div></li><li><div class="item-name"><a class="struct" href="struct.LoopControl.html" title="struct spirv::LoopControl">LoopControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_loop_control_a_loop_control">LoopControl</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemoryAccess.html" title="struct spirv::MemoryAccess">MemoryAccess</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_access_a_memory_access">MemoryAccess</a></div></li><li><div class="item-name"><a class="struct" href="struct.MemorySemantics.html" title="struct spirv::MemorySemantics">MemorySemantics</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_semantics_a_memory_semantics">MemorySemantics</a></div></li><li><div class="item-name"><a class="struct" href="struct.RayFlags.html" title="struct spirv::RayFlags">RayFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_flags_a_ray_flags">RayFlags</a></div></li><li><div class="item-name"><a class="struct" href="struct.SelectionControl.html" title="struct spirv::SelectionControl">SelectionControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_selection_control_a_selection_control">SelectionControl</a></div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.AccessQualifier.html" title="enum spirv::AccessQualifier">AccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_access_qualifier_a_access_qualifier">AccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.AddressingModel.html" title="enum spirv::AddressingModel">AddressingModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_addressing_model_a_addressing_model">AddressingModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.BuiltIn.html" title="enum spirv::BuiltIn">BuiltIn</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_built_in_a_built_in">BuiltIn</a></div></li><li><div class="item-name"><a class="enum" href="enum.CLOp.html" title="enum spirv::CLOp">CLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/OpenCL.ExtendedInstructionSet.100.html">OpenCL.std</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.Capability.html" title="enum spirv::Capability">Capability</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_capability_a_capability">Capability</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixLayout.html" title="enum spirv::CooperativeMatrixLayout">CooperativeMatrixLayout</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_layout_a_cooperative_matrix_layout">CooperativeMatrixLayout</a></div></li><li><div class="item-name"><a class="enum" href="enum.CooperativeMatrixUse.html" title="enum spirv::CooperativeMatrixUse">CooperativeMatrixUse</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_cooperative_matrix_use_a_cooperative_matrix_use">CooperativeMatrixUse</a></div></li><li><div class="item-name"><a class="enum" href="enum.Decoration.html" title="enum spirv::Decoration">Decoration</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_decoration_a_decoration">Decoration</a></div></li><li><div class="item-name"><a class="enum" href="enum.Dim.html" title="enum spirv::Dim">Dim</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_dim_a_dim">Dim</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionMode.html" title="enum spirv::ExecutionMode">ExecutionMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_mode_a_execution_mode">ExecutionMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.ExecutionModel.html" title="enum spirv::ExecutionModel">ExecutionModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_execution_model_a_execution_model">ExecutionModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPDenormMode.html" title="enum spirv::FPDenormMode">FPDenormMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_denorm_mode_a_fp_denorm_mode">FPDenormMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPOperationMode.html" title="enum spirv::FPOperationMode">FPOperationMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_operation_mode_a_fp_operation_mode">FPOperationMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FPRoundingMode.html" title="enum spirv::FPRoundingMode">FPRoundingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_fp_rounding_mode_a_fp_rounding_mode">FPRoundingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.FunctionParameterAttribute.html" title="enum spirv::FunctionParameterAttribute">FunctionParameterAttribute</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_function_parameter_attribute_a_function_parameter_attribute">FunctionParameterAttribute</a></div></li><li><div class="item-name"><a class="enum" href="enum.GLOp.html" title="enum spirv::GLOp">GLOp</a></div><div class="desc docblock-short"><a href="https://www.khronos.org/registry/spir-v/specs/unified1/GLSL.std.450.html">GLSL.std.450</a> extended instruction opcode</div></li><li><div class="item-name"><a class="enum" href="enum.GroupOperation.html" title="enum spirv::GroupOperation">GroupOperation</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_group_operation_a_group_operation">GroupOperation</a></div></li><li><div class="item-name"><a class="enum" href="enum.HostAccessQualifier.html" title="enum spirv::HostAccessQualifier">HostAccessQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_host_access_qualifier_a_host_access_qualifier">HostAccessQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelDataType.html" title="enum spirv::ImageChannelDataType">ImageChannelDataType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_data_type_a_image_channel_data_type">ImageChannelDataType</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageChannelOrder.html" title="enum spirv::ImageChannelOrder">ImageChannelOrder</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_channel_order_a_image_channel_order">ImageChannelOrder</a></div></li><li><div class="item-name"><a class="enum" href="enum.ImageFormat.html" title="enum spirv::ImageFormat">ImageFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_format_a_image_format">ImageFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.InitializationModeQualifier.html" title="enum spirv::InitializationModeQualifier">InitializationModeQualifier</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_initialization_mode_qualifier_a_initialization_mode_qualifier">InitializationModeQualifier</a></div></li><li><div class="item-name"><a class="enum" href="enum.KernelEnqueueFlags.html" title="enum spirv::KernelEnqueueFlags">KernelEnqueueFlags</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_kernel_enqueue_flags_a_kernel_enqueue_flags">KernelEnqueueFlags</a></div></li><li><div class="item-name"><a class="enum" href="enum.LinkageType.html" title="enum spirv::LinkageType">LinkageType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_linkage_type_a_linkage_type">LinkageType</a></div></li><li><div class="item-name"><a class="enum" href="enum.LoadCacheControl.html" title="enum spirv::LoadCacheControl">LoadCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_load_cache_control_a_load_cache_control">LoadCacheControl</a></div></li><li><div class="item-name"><a class="enum" href="enum.MemoryModel.html" title="enum spirv::MemoryModel">MemoryModel</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_memory_model_a_memory_model">MemoryModel</a></div></li><li><div class="item-name"><a class="enum" href="enum.Op.html" title="enum spirv::Op">Op</a></div><div class="desc docblock-short">SPIR-V <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_instructions_a_instructions">instructions</a> opcodes</div></li><li><div class="item-name"><a class="enum" href="enum.OverflowModes.html" title="enum spirv::OverflowModes">OverflowModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_overflow_modes_a_overflow_modes">OverflowModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.PackedVectorFormat.html" title="enum spirv::PackedVectorFormat">PackedVectorFormat</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_packed_vector_format_a_packed_vector_format">PackedVectorFormat</a></div></li><li><div class="item-name"><a class="enum" href="enum.QuantizationModes.html" title="enum spirv::QuantizationModes">QuantizationModes</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_quantization_modes_a_quantization_modes">QuantizationModes</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCandidateIntersectionType.html" title="enum spirv::RayQueryCandidateIntersectionType">RayQueryCandidateIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_candidate_intersection_type_a_ray_query_candidate_intersection_type">RayQueryCandidateIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryCommittedIntersectionType.html" title="enum spirv::RayQueryCommittedIntersectionType">RayQueryCommittedIntersectionType</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_committed_intersection_type_a_ray_query_committed_intersection_type">RayQueryCommittedIntersectionType</a></div></li><li><div class="item-name"><a class="enum" href="enum.RayQueryIntersection.html" title="enum spirv::RayQueryIntersection">RayQueryIntersection</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_ray_query_intersection_a_ray_query_intersection">RayQueryIntersection</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerAddressingMode.html" title="enum spirv::SamplerAddressingMode">SamplerAddressingMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_addressing_mode_a_sampler_addressing_mode">SamplerAddressingMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.SamplerFilterMode.html" title="enum spirv::SamplerFilterMode">SamplerFilterMode</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_sampler_filter_mode_a_sampler_filter_mode">SamplerFilterMode</a></div></li><li><div class="item-name"><a class="enum" href="enum.Scope.html" title="enum spirv::Scope">Scope</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_scope_a_scope">Scope</a></div></li><li><div class="item-name"><a class="enum" href="enum.SourceLanguage.html" title="enum spirv::SourceLanguage">SourceLanguage</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_source_language_a_source_language">SourceLanguage</a></div></li><li><div class="item-name"><a class="enum" href="enum.StorageClass.html" title="enum spirv::StorageClass">StorageClass</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_storage_class_a_storage_class">StorageClass</a></div></li><li><div class="item-name"><a class="enum" href="enum.StoreCacheControl.html" title="enum spirv::StoreCacheControl">StoreCacheControl</a></div><div class="desc docblock-short">SPIR-V operand kind: <a href="https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_store_cache_control_a_store_cache_control">StoreCacheControl</a></div></li></ul><h2 id="constants" class="section-header">Constants<a href="#constants" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="constant" href="constant.MAGIC_NUMBER.html" title="constant spirv::MAGIC_NUMBER">MAGIC_NUMBER</a></div></li><li><div class="item-name"><a class="constant" href="constant.MAJOR_VERSION.html" title="constant spirv::MAJOR_VERSION">MAJOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.MINOR_VERSION.html" title="constant spirv::MINOR_VERSION">MINOR_VERSION</a></div></li><li><div class="item-name"><a class="constant" href="constant.REVISION.html" title="constant spirv::REVISION">REVISION</a></div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.Word.html" title="type spirv::Word">Word</a></div></li></ul></section></div></main></body></html>