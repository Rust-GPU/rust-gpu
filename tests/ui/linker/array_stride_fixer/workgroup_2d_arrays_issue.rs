// Test that reproduces the OpInBoundsAccessChain type mismatch issue
// with workgroup 2D arrays after array_stride_fixer changes

// build-pass
// compile-flags: -C llvm-args=--disassemble-globals
// only-vulkan1.1
// normalize-stderr-test "OpLine .*\n" -> ""
// normalize-stderr-test "OpSource .*\n" -> ""
// normalize-stderr-test "\S*/lib/rustlib/" -> "$SYSROOT/lib/rustlib/"

use spirv_std::spirv;

const TILE_SIZE: u32 = 32;

#[spirv(compute(threads(32, 32)))]
pub fn transpose_2d_workgroup(
    #[spirv(local_invocation_id)] lid: spirv_std::glam::UVec3,
    #[spirv(workgroup)] shared_real: &mut [[f32; TILE_SIZE as usize]; TILE_SIZE as usize],
    #[spirv(workgroup)] shared_imag: &mut [[f32; TILE_SIZE as usize]; TILE_SIZE as usize],
) {
    let lx = lid.x as usize;
    let ly = lid.y as usize;

    // This should trigger the OpInBoundsAccessChain issue
    shared_real[ly][lx] = 1.0;
    shared_imag[ly][lx] = 2.0;

    // Read back to ensure usage
    let _val = shared_real[lx][ly] + shared_imag[lx][ly];
}
