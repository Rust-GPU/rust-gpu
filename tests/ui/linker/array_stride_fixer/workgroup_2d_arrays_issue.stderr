OpCapability Shader
OpCapability Float64
OpCapability Int64
OpCapability Int16
OpCapability Int8
OpCapability ShaderClockKHR
OpCapability VulkanMemoryModel
OpExtension "SPV_KHR_shader_clock"
OpExtension "SPV_KHR_vulkan_memory_model"
OpMemoryModel Logical Vulkan
OpEntryPoint GLCompute %1 "transpose_2d_workgroup" %2
OpExecutionMode %1 LocalSize 32 32 1
%3 = OpString "$OPSTRING_FILENAME/workgroup_2d_arrays_issue.rs"
OpName %4 "shared_real"
OpName %5 "shared_imag"
OpDecorate %2 BuiltIn LocalInvocationId
%6 = OpTypeInt 32 0
%7 = OpTypeVector %6 3
%8 = OpTypePointer Input %7
%9 = OpTypeVoid
%10 = OpTypeFunction %9
%2 = OpVariable  %8  Input
%11 = OpTypeBool
%12 = OpConstant  %6  32
%13 = OpTypeFloat 32
%14 = OpTypeArray %13 %12
%15 = OpTypePointer Workgroup %14
%16 = OpTypeArray %14 %12
%17 = OpTypePointer Workgroup %16
%4 = OpVariable  %17  Workgroup
%18 = OpTypePointer Workgroup %13
%19 = OpConstant  %13  1065353216
%5 = OpVariable  %17  Workgroup
%20 = OpConstant  %13  1073741824
